### PR TITLE
Add {RO}Span GetReference and ROMemory TryGetArray to MemoryMarshal

### PR DESCRIPTION
**Pull request by ahsonkhan**
Part of:
https://github.com/dotnet/corefx/issues/25412
https://github.com/dotnet/corefx/issues/25615

Depends on the PR in coreclr - https://github.com/dotnet/coreclr/pull/15417

cc @jkotas, @stephentoub, @KrzysztofCwalina

### CODE REVIEW

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

**Comment by ahsonkhan** [comment]
Is this PR good to merge?

File: src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 1]
```diff
@@ -23,5 +24,54 @@ public static class MemoryMarshal
...
+        /// If unable to get the array segment, return false with a default array segment.
+        /// </summary>
+        public static bool TryGetArray<T>(ReadOnlyMemory<T> readOnlyMemory, out ArraySegment<T> arraySegment)
+        {
+            object obj = readOnlyMemory.GetObjectStartLength(out int index, out int length);
+            if (index < 0)
+            {
+                if (((OwnedMemory<T>)obj).TryGetArray(out var segment))
+                {
+                    arraySegment = new ArraySegment<T>(segment.Array, segment.Offset + (index & ReadOnlyMemory<T>.RemoveOwnedFlagBitMask), length);
```
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.cs]: Do we need to check that the length <= segment.Length?

Comment (ahsonkhan) [reply src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.cs]: If length is greater than segment.Count (if, let's say someone implements OwnedMemory.TryGetArray incorrectly and returns a subset of the backing memory), then the ArraySegment constructor will already throw ArgumentException.


```text
System.ArgumentException : Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection.
```

Regarding index, that only changes on a Slice operation, and we decrease the length accordingly whenever that happens. Therefore, I don't think an explicit check is necessary here.

**Pull request merged**